perf_benchmark/benchmark_input_stream.cc

```diff
@@ -0,0 +1,160 @@
+    : BenchmarkZeroCopyInputStream(std::move(msg), chunk_per_msg), pos_(0) {}
+
+int64_t UnaryBenchmarkZeroCopyInputStream::BytesAvailable() const {
+  if (finished_) {
+    return 0;
+  }
+  // check if we are at the last chunk
+  if (pos_ + chunk_size_ >= msg_.size()) {
+    return msg_.size() - pos_;
+  }
+  return chunk_size_;
+}
+
+bool UnaryBenchmarkZeroCopyInputStream::Next(const void** data, int* size) {
+  if (finished_) {
+    *size = 0;
+    return false;
+  }
+  *data = msg_.data() + pos_;
+  if (pos_ + chunk_size_ >= msg_.size()) { // last message
+    *size = msg_.size() - pos_;
+    pos_ = 0; // reset pos after sending the last message
+    finished_ = true;
+  } else {
+    *size = chunk_size_;
+    pos_ += chunk_size_;
+  }
+  return true;
+}
+
+void UnaryBenchmarkZeroCopyInputStream::Reset() {
+  finished_ = false;
+}
+
+uint64_t UnaryBenchmarkZeroCopyInputStream::TotalBytes() const {
+  return msg_.size();
+}
+
+StreamingBenchmarkZeroCopyInputStream::StreamingBenchmarkZeroCopyInputStream(std::string msg,
+                                                                             uint64_t chunk_per_msg,
+                                                                             uint64_t stream_size)
+    : BenchmarkZeroCopyInputStream(std::move(msg), chunk_per_msg),
+      pos_(0),
+      stream_size_(stream_size),
+      msg_sent(0) {
+  if (stream_size_ == 1) {
+    // Edge case, we only need to set header_
+    header_ = "[" + msg_ + "]";
+  } else {
+    header_ = "[" + msg_ + ", ";
+    body_ = msg_ + ", ";
+    tail_ = msg_ + "]";
+  }
+}
+int64_t StreamingBenchmarkZeroCopyInputStream::BytesAvailable() const {
+  // no message sent -> next message is header_
+  if (msg_sent == 0) {
+    // header has 3 chars overhead
+    if (pos_ + chunk_size_ + 3 >= header_.size()) {
+      return header_.size() - pos_;
+    }
+    return chunk_size_;
+  }
+  // last message to be sent -> next message is tail_
+  if (msg_sent + 1 == stream_size_) {
+    // tail has 1 char overhead
+    if (pos_ + chunk_size_ + 1 >= tail_.size()) {
+      return tail_.size() - pos_;
+    }
+    return chunk_size_;
+  }
+  // otherwise -> next message is body_
+  // body has 2 chars overhead
+  if (pos_ + chunk_size_ + 2 >= body_.size()) {
+    return body_.size() - pos_;
+  }
+  return chunk_size_;
+}
+bool StreamingBenchmarkZeroCopyInputStream::Next(const void** data, int* size) {
+  std::string* msg_ptr;
+  int overhead = 0; // character overhead due to [,] characters.
+  if (msg_sent == 0) {
+    // no message sent -> next message is header_
+    msg_ptr = &header_;
+    overhead = 3;
+  } else if (msg_sent + 1 == stream_size_) {
+    // last message to be sent -> next message is tail_
+    msg_ptr = &tail_;
+    overhead = 1;
+  } else {
+    msg_ptr = &body_;
+    overhead = 2;
+  }
+
+  *data = msg_ptr->data() + pos_;
+  if (pos_ + chunk_size_ + overhead >= msg_ptr->size()) { // last message
```

Comment:
Did not get why "overhead" need to be here.  msg_ptr->size() should already have "overhead"

e.g.  if msg_.size() is 10,   header_.size() is 12 in both cases,  msg_ptr->size() is 12, why we need to add overhead=3 here?

Reply:
Yes, I think you are right. We shouldn't need the `overhead` here. That was a great catch :) 
I just writing the unit test for this method. I will fix this along the way.